File: api/generate.rs
// api/generate.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::env;
use vercel_runtime::{run, Body, Error, Request, Response, StatusCode};

#[derive(Deserialize)]
struct GenerateRequest {
    prompt: String,
    model: Option<String>,
    max_tokens: Option<u32>,
    temperature: Option<f32>,
    system: Option<String>,
}

#[derive(Serialize, Deserialize)]
struct AnthropicResponse {
    content: Option<Vec<AnthropicContent>>,
}

#[derive(Serialize, Deserialize)]
struct AnthropicContent {
    text: String,
}

#[tokio::main]
async fn main() -> Result<(), Error> {
    run(handler).await
}

pub async fn handler(req: Request) -> Result<Response<Body>, Error> {
    // Handle CORS preflight requests
    if req.method() == "OPTIONS" {
        return Response::builder()
            .status(StatusCode::OK)
            .header("Access-Control-Allow-Origin", "*")
            .header("Access-Control-Allow-Methods", "POST, OPTIONS")
            .header("Access-Control-Allow-Headers", "Content-Type")
            .body("".into())
            .map_err(|e| Error::from(format!("Failed to build CORS response: {}", e)));
    }

    // Only allow POST requests
    if req.method() != "POST" {
        return Response::builder()
            .status(405)
            .header("Allow", "POST, OPTIONS")
            .body("Method Not Allowed".into())
            .map_err(|e| Error::from(format!("Failed to build response: {}", e)));
    }

    // Parse request body
    let body = req.body();
    let payload: GenerateRequest = serde_json::from_slice(body)
        .map_err(|e| Error::from(format!("Failed to parse request body: {}", e)))?;
    
    let client = Client::new();
    
    // Get API key from environment
    let api_key = env::var("ANTHROPIC_API_KEY")
        .map_err(|_| Error::from("ANTHROPIC_API_KEY environment variable is required"))?;
    
    // Build the request body with configurable parameters
    let mut request_body = serde_json::json!({
        "model": payload.model.unwrap_or_else(|| "claude-3-haiku-20240307".to_string()),
        "max_tokens": payload.max_tokens.unwrap_or(4096),
        "temperature": payload.temperature.unwrap_or(1.0),
        "messages": [{ 
            "role": "user", 
            "content": [{ 
                "type": "text", 
                "text": payload.prompt 
            }] 
        }],
    });

    // Add system message if provided
    if let Some(system_msg) = payload.system {
        request_body["system"] = serde_json::json!(system_msg);
    }

    // Make request to Anthropic API
    let resp = client
        .post("https://api.anthropic.com/v1/messages")
        .header("Content-Type", "application/json")
        .header("x-api-key", api_key)
        .header("anthropic-version", "2023-06-01")
        .json(&request_body)
        .send()
        .await
        .map_err(|e| Error::from(format!("Request failed: {}", e)))?;

    // Check if the response status indicates an error
    if !resp.status().is_success() {
        let status = resp.status();
        let error_text = resp.text().await.unwrap_or_else(|_| "Unknown error".to_string());
        return Err(Error::from(format!("Anthropic API error: {} - {}", status, error_text)));
    }

    let data: AnthropicResponse = resp
        .json()
        .await
        .map_err(|e| Error::from(format!("Failed to parse response: {}", e)))?;

    if let Some(content) = data.content.and_then(|mut c| c.pop()) {
        let json_body = serde_json::to_string(&content)
            .map_err(|e| Error::from(format!("Failed to serialize response: {}", e)))?;
        
        Response::builder()
            .status(StatusCode::OK)
            .header("Content-Type", "application/json")
            .header("Access-Control-Allow-Origin", "*")
            .header("Access-Control-Allow-Methods", "POST, OPTIONS")
            .header("Access-Control-Allow-Headers", "Content-Type")
            .body(json_body.into())
            .map_err(|e| Error::from(format!("Failed to build response: {}", e)))
    } else {
        Err(Error::from("No content returned from API"))
    }
}

File: api/generate.ts
// Using direct API calls instead of WASM for now
// WASM integration can be added later once Vercel deployment is working

const systemMessage = `You are an HTML App Generator that creates complete, interactive web applications as self-contained HTML code. Your output will be inserted directly into a React component using dangerouslySetInnerHTML, so you must generate valid, safe HTML that works immediately when rendered.

## CRITICAL REQUIREMENTS

### HTML Output Format
- Generate ONLY complete, valid HTML markup
- NO markdown, NO code blocks, NO explanations
- Start directly with HTML tags (e.g., <div>, <html>, etc.)
- All content must be self-contained within the generated HTML

### Safety Requirements
- Use only safe HTML elements and attributes
- Avoid potentially dangerous elements like <iframe>, <object>, <embed>
- No external script sources or imports
- All JavaScript must be inline within <script> tags
- All CSS must be inline within <style> tags or as style attributes

### Interactivity Guidelines
- Use inline JavaScript event handlers (onclick, onchange, etc.)
- Implement functionality with vanilla JavaScript within <script> tags
- Create interactive elements like buttons, forms, modals, tabs, etc.
- Use modern JavaScript features (ES6+, async/await, fetch API)
- Implement local storage for data persistence when appropriate

### Styling Requirements
- Use inline styles, <style> tags, or modern CSS within the HTML
- Implement responsive designs with CSS Grid, Flexbox, or media queries
- Use modern CSS features (CSS Variables, animations, transitions)
- Create visually appealing interfaces with proper colors, spacing, typography
- Ensure accessibility with proper contrast and semantic HTML

### Application Types You Can Create
- Productivity tools (todo lists, note-takers, calculators)
- Games (puzzles, card games, simple arcade games)
- Utilities (converters, generators, timers)
- Educational tools (quizzes, flashcards, tutorials)
- Creative tools (drawing apps, text editors, color pickers)
- Data visualization tools (charts, dashboards)
- Small business tools (forms, surveys, simple CRM)

### Technical Capabilities
- Local data storage using localStorage or sessionStorage
- API calls using fetch() for external data (when CORS allows)
- File handling using File API (for local file processing)
- Canvas API for graphics and drawing
- Web APIs (Geolocation, Notifications, etc. when appropriate)
- CSS animations and transitions for smooth UX

### Code Organization
- Structure HTML semantically with proper headings, sections, etc.
- Group related JavaScript functions together
- Organize CSS logically (reset, layout, components, utilities)
- Use comments to explain complex logic
- Keep code readable and maintainable

### Example Response Structure
When asked to create an app, respond with complete HTML like this:

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>App Name</title>
    <style>
        /* Your CSS here */
    </style>
</head>
<body>
    <!-- Your HTML content here -->
    
    <script>
        // Your JavaScript here
    </script>
</body>
</html>

### Quality Standards
- Create fully functional applications, not just mockups
- Ensure all features work as expected
- Implement proper error handling
- Make interfaces intuitive and user-friendly
- Test edge cases mentally before generating code
- Optimize for both desktop and mobile experiences

### What NOT to Include
- External dependencies or CDN links
- Server-side code or backend requirements  
- Build processes or compilation steps
- Package managers or npm dependencies
- Framework-specific code (React, Vue, Angular components)
- Explanatory text outside of HTML comments

Remember: You are creating complete, ready-to-use web applications that work immediately when inserted into a web page. Focus on functionality, usability, and clean code that runs in any modern browser.`;

export default async function handler(req, res) {
  // Handle CORS
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");

  if (req.method === "OPTIONS") {
    return res.status(200).end();
  }

  if (req.method !== "POST") {
    return res.status(405).json({ error: "Method not allowed" });
  }

  try {
    const { prompt } = req.body;

    const apiKey = process.env.ANTHROPIC_API_KEY;
    if (!apiKey) {
      return res
        .status(500)
        .json({ error: "ANTHROPIC_API_KEY environment variable is required" });
    }

    // Build the request body
    const requestBody = {
      model: "claude-3-haiku-20240307",
      max_tokens: 4096,
      temperature: 1.0,
      system: systemMessage,
      messages: [
        {
          role: "user",
          content: [
            {
              type: "text",
              text: prompt,
            },
          ],
        },
        {
          role: "assistant",
          content: [
            {
              type: "text",
              text: "<",
            },
          ],
        },
      ],
    };

    // Make request to Anthropic API
    const response = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-api-key": apiKey,
        "anthropic-version": "2023-06-01",
      },
      body: JSON.stringify(requestBody),
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`Anthropic API error: ${response.status} - ${errorText}`);
      return res
        .status(response.status)
        .json({ error: `API error: ${response.status}` });
    }

    const data = await response.json();

    if (data.content && data.content.length > 0) {
      const content = data.content[0];
      return res.status(200).json(content);
    } else {
      return res.status(500).json({ error: "No content returned from API" });
    }
  } catch (error) {
    console.error("Error in generate endpoint:", error);
    return res.status(500).json({ error: error.message || "Unknown error" });
  }
}
